readme.md
# Phockup

[![Tests](https://github.com/ivandokov/phockup/actions/workflows/tests.yml/badge.svg)](https://github.com/ivandokov/phockup/actions/workflows/tests.yml)
[![Deploy](https://github.com/ivandokov/phockup/actions/workflows/deploy.yml/badge.svg)](https://github.com/ivandokov/phockup/actions/workflows/deploy.yml)
[![Lint](https://github.com/ivandokov/phockup/actions/workflows/lint.yml/badge.svg)](https://github.com/ivandokov/phockup/actions/workflows/lint.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](license)

Media sorting tool to organize photos and videos from your camera in folders by year, month and day.

## How it works
The software will collect all files from the input directory and copy them to the output directory without changing the files content. It will only rename the files and place them in the proper directory for year, month and day.

All files which are not images or videos or those which do not have creation date information will be placed in a directory called `unknown` without file name change. By doing this you can be sure that the input directory can be safely deleted after the successful process completion because **all** files from the input directory have a copy in the output directory.

If the target file already exists, its checksum is compared with the source to determine if it is a duplicate. If the checksums are different, we do not have a duplicate and the target filename will be suffixed with a number, for example "-1". If the checksums match, the copy operation will be skipped.

## Installation

### Docker

The docker container supports two operation modes. The first allows for a single execution of phockup. In this mode, the container will be stopped after the execution is complete. The second mode allows for execution in intervals. In this mode, the container will continue running until the user decides to stop it.

#### Single execution mode
In this mode, all phockup parameters need to be passed as direct parameters within the docker run command. As you define a complete set of phockup parameters for this execution mode, this includes the paths to the input and output folders within the container.
To execute phockup only once, use the following command:

```
docker run -v ~/Pictures:/mnt ivandokov/phockup:latest /mnt/input /mnt/output [PHOCKUP ARGUMENTS]
```

#### Continuous execution mode
In this mode, all relevant settings are defined through environment variables and volume mappings. The folders where phockup moves files are always /mnt/input and /mnt/output within the container and can not be changed. You can of course map any folder on your host system to those folders within the container.

The `-v ~/Pictures/input:/mnt/input` part of the command mounts your `~/Pictures/input` directory to `/mnt/input` inside the container. The same is done for the output folder. You can pass any **absolute** path to be mounted to the container and later on be used as paths for the `phockup` command. The example above provides your `~/Pictures/input` as `INPUTDIR` and `~/Pictures/output` as `OUTPUDIR`. You can pass additional arguments through the `OPTIONS` environment variable.

To keep the container running and execute phockup in intervals, use the following command:

```
docker run -v ~/Pictures/input:/mnt/input -v ~/Pictures/output:/mnt/output -e "CRON=* * * * *" -e "OPTIONS=[PHOCKUP ARGUMENTS]" ivandokov/phockup:latest
```

This will execute phockup once every minute (as defined by the [value of the CRON environment variable](https://crontab.guru/#*_*_*_*_*)). However, the container will not spawn a new phockup process if another phockup process is still running. You can define other intervals for execution using the usual cron syntax. If you want to pass further arguments to phockup, use the OPTIONS environment variable. In this execution mode, phockup will always use the directories mounted to `/mnt/input` and `/mnt/output` and ignore arguments passed in the style of the single execution mode.

### Mac
Requires [Homebrew](http://brew.sh/)
```
brew tap ivandokov/homebrew-contrib
brew install phockup
```

### Linux (snap)
Requires [snapd](https://snapcraft.io/docs/core/install)
```
sudo snap install phockup
```
*Note: snap applications can access files only in your **home and `/media` directories** for security reasons. If your media files are not located in these directories you should use the installation method below.
If your files are in `/media` you should run the following command to allow access:*
```
sudo snap connect phockup:removable-media
```

### Linux (without snap)
If you are using distro which doesn't support [snapd](https://snapcraft.io/docs/core/install) or you don't want to download the snap you can use the following commands to download the source and set it up
```
sudo apt-get install python3 libimage-exiftool-perl -y
curl -L https://github.com/ivandokov/phockup/archive/latest.tar.gz -o phockup.tar.gz
tar -zxf phockup.tar.gz
sudo mv phockup-* /opt/phockup
cd /opt/phockup
pip3 install -r requirements.txt
sudo ln -s /opt/phockup/phockup.py /usr/local/bin/phockup
```

### Linux (AUR)

If you are an arch user you can install from the [aur](https://aur.archlinux.org/packages/phockup).

For example using [yay](https://github.com/Jguer/yay):

```bash
yay -S phockup
```

### Windows
* Download and install latest stable [Python 3](https://www.python.org/downloads/windows/)
* Download Phockup's [latest release](https://github.com/ivandokov/phockup/archive/latest.tar.gz) and extract the archive
* Download exiftool from the official [website](https://exiftool.org/) and extract the archive
* Rename `exiftool(-k).exe` to `exiftool.exe`
* Move `exiftool.exe` to phockup folder
* Open Command Prompt and `cd` to phockup folder
* Use the command below (use `phockup.py` instead of `phockup`)

## Usage
Organize photos from one directory into another
```
phockup INPUTDIR OUTPUTDIR
```

`INPUTDIR` is the directory where your photos are located.
`OUTPUTDIR` is the directory where your **sorted** photos will be stored. It could be a new not existing directory.

Example:
```
phockup ~/Pictures/camera ~/Pictures/sorted
```

### Version
If you want to view the version of phockup use the flag `-v | --version`.

### Date format
If you want to change the output directories date format you can do it by passing the format as `-d | --date` argument.
You can choose different year format (e.g. 17 instead of 2017) or decide
        to skip the day directories and have all photos sorted in year/month.

```
Supported formats:
    YYYY - 2016, 2017 ...
    YY   - 16, 17 ...
    MM   - 07, 08, 09 ...
    M    - July, August, September ...
    m    - Jul, Aug, Sept ...
    DD   - 27, 28, 29 ... (day of month)
    DDD  - 123, 158, 365 ... (day of year)
    U    - 00, 01, 53 ... (week of the year, Sunday first day of week)
    W    - 00, 01, 53 ... (week of the year, Monday first day of week)

Example:
    YYYY/MM/DD -> 2011/07/17
    YYYY/M/DD  -> 2011/July/17
    YYYY/m/DD  -> 2011/Jul/17
    YY/m-DD    -> 11/Jul-17
    YYYY/U     -> 2011/30
    YYYY/W     -> 2011/28
```

### Prefix/Suffix
In order to support both aggregation and finer granularity of files
sorted, you can specify a prefix or suffix (or both) to aid in storing
files in directories beyond strictly date.

*NOTE:* Prefixes and suffixes will also apply to the **'unknown'** folder to
isolate files that cannot be processed into their respective folders.
This creates a bit more chaos for 'unknown' files, but should allow
them to be managed by whomever they "belong" to.

#### Prefix
`--output-prefix` flag can be used to specify a directory to be
appended to the `OUTPUTDIR`, and thus prepended to the date.

For example:
```
phockup ~/Pictures/camera /mnt/sorted --output_prefix=nikon
```
would place files in folders similar to:
```
/mnt/sorted/nikon/2011/07/17
/mnt/sorted/nikon/unknown
```

While it may seem to be redundant with `OUTPUTDIR`, this flag is
intended to add support for more cleanly determining the output
directory at run-time via environment variable expansion (i.e. use
$USER, %USERNAME%, $HOSTNAME, etc. to aggregate files)

For example:
```
phockup ~/Pictures/camera /mnt/sorted --output_prefix=$USER
```

would yield an output directory of
```
/mnt/sorted/ivandokov/2011/07/17
/mnt/sorted/ivandokov/unknown
```

This allows the same script to be deployed to multiple users/machines
and allows sorting into their respective top level directories.

#### Suffix
`--output-suffix` flag can be used to specify a directory within the
target date directory for a file.  This allows files to be sorted in
their respective date/time folders while additionally adding a
directory based on the suffix value for additional metadata.

For example:
```
phockup ~/Pictures/DCIM/NIKOND40 /mnt/sorted --output_suffix=nikon
phockup ~/Pictures/DCIM/100APPLE /mnt/sorted --output_suffix=iphone
```

This would allow files to be stored in the following structure:

```
/mnt/sorted/2011/07/17/nikon/DCS_0001.NEF
...
/mnt/sorted/2011/07/17/nikon/DCS_0099.NEF
/mnt/sorted/unknown/nikon/

/mnt/sorted/2011/07/17/iphone/ABIL6163.HEIC
...
/mnt/sorted/2011/07/17/iphone/YZYE9497.HEIC
/mnt/sorted/unknown/iphone/
```

The output suffix also allows for environment variable expansion (e.g.
$USER, $HOSTNAME, %USERNAME%, etc.) allowing dynamic folders to
represent additional metadata about the images.

For example:

```
phockup ~/Pictures/ /mnt/sorted --output_suffix=$HOSTNAME

or

phockup ~/Pictures/ /mnt/sorted --output_suffix=$USER
```
could be used to sort images based on the source computer or user,
perventing hetrogenous collections of images from disparate sources
saving to the same central respository.

The two options above can be used to help sort/store images

#### Limit files processed by date
`--from-date` flag can be used to limit the operations to the files that are newer than the provided date (inclusive).
The date must be specified in format YYYY-MM-DD. Files with unknown date won't be skipped.

For example:
```
phockup ~/Pictures/DCIM/NIKOND40 ~/Pictures/sorted --from-date="2017-01-02"
```
`--to-date` flag can be used to limit the operations to the files that are older than the provided date (inclusive).
The date must be specified in format YYYY-MM-DD. Files with unknown date won't be skipped.

For example:
```
phockup ~/Pictures/DCIM/NIKOND40 ~/Pictures/sorted --to-date="2017-01-02"
```

`--from-date` and `--to-date` can be combined for better control over the files that are processed.

For example:
```
phockup ~/Pictures/DCIM/NIKOND40 ~/Pictures/sorted --from-date="2017-01-02" --to-date="2017-01-03"
```

### Missing date information in EXIF
If any of the photos does not have date information you can use the `-r | --regex` option to specify date format for date extraction from filenames:
```
--regex="(?P<day>\d{2})\.(?P<month>\d{2})\.(?P<year>\d{4})[_-]?(?P<hour>\d{2})\.(?P<minute>\d{2})\.(?P<second>\d{2})"
```

As a last resort, specify the `-t | --timestamp` option to use the file modification timestamp. This may not be accurate in all cases but can provide some kind of date if you'd rather it not go into the `unknown` folder.

### Move files
Instead of copying the process will move all files from the INPUTDIR to the OUTPUTDIR by using the flag `-m | --move`. This is useful when working with a big collection of files and the remaining free space is not enough to make a copy of the INPUTDIR.

### Link files
Instead of copying the process will create hard link all files from the INPUTDIR into new structure in OUTPUTDIR by using the flag `-l | --link`. This is useful when working with good structure of photos in INPUTDIR (like folders per device).

### Original filenames
Organize the files in selected format or using the default year/month/day format but keep original filenames by using the flag `-o | --original-names`.

### File Type
By default, Phockup addresses both image and video files. If you want to restrict your command to either images or videos only, use `--file-type=[image|video]`.

### Fix incorrect dates
If date extracted from photos is incorrect, you can use the `-f | --date-field` option to set the correct exif field to get date information from. Use this command to list which fields are available for a file:
```
exiftool -time:all -mimetype -j file.jpg
```
The output may look like this, but with more fields:
```
[{
  "DateTimeOriginal": "2017:10:06 01:01:01",
  "CreateDate": "2017:01:01 01:01:01",
]}
```
If the correct date is in `DateTimeOriginal`, you can include the option `--date-field=DateTimeOriginal` to get date information from it.
To set multiple fields to be tried in order until a valid date is found, just join them with spaces in a quoted string like `"CreateDate FileModifyDate"`.

### Handle sidecars
Sidecars are files that have the same name as an image or video file, but a different extension. They typically contain additional metadata pertaining to the image or video file.

File extensions which are considered to be sidecars if they are named the same as a corresponding image file are:

 * .xmp
 * .json
 * .yaml
 * .yml

So, for example, if `image.jpg` exists, then `image.xmp` (or `image.jpg.xmp`) will be considered a sidecar file of image.jpg.

When moving the main file, sidecars will inherit the name of the main file and be moved to the same location rather than being placed in the unknown directory.

You can change which file extensions are eligible to be considered sidecars using the `--sidecars` argument. For example, to only treat `.xmp` and `.json` as sidecars:

```
--sidecars='xmp,json'
```

To disable handling sidecars entirely:

```
--sidecars=''
```


### Dry run
If you want phockup to run without any changes (don't copy/move any files) but just show which changes would be done, enable this feature by using the flag `-y | --dry-run`.

### Log
If you want phockup to run and store the output in a log file use the flag `--log`. This flag can be used in conjunction with the flags `--quiet` or `--progress`.
```
--log=<PATH>/log.txt
```

### Quiet run
If you want phockup to run without any output (displaying only error messages, and muting all progress messages) use the flag `--quiet`.

### Progress run
If you want phockup to run with a progressbar (displaying only the progress and muting all progress messages (including errors)) use the flag `--progress`.


### Limit directory traversal depth
If you would like to limit how deep the directories are traversed, you can use the `--maxdepth` option to specify the maximum number of levels below the input directory to process.  In order to process only the input directory, you can disable sub-directory processing with:
`--maxdepth=0`  The current implementation is limited to a maximum depth of 255.

### Improving throughput with concurrency
If you want to allocate additional CPUs/cores to the image processing
operations, you can specify additional resources via the
`--max-concurrency` flag. Specifying `--max-concurrency=n`, where `n`
represents the maximum number of operations to attempt
concurrently, will leverage the additional CPU resources to start
additional file operations while waiting for file I/O.  This can lead
to significant increases in file processing throughput.

Due to how concurrency is implemented in Phockup (specifically
`ThreadPoolExecutor`), this option has the greatest impact on
directories with a large numbers of files in them,
versus many directories with small numbers of files in each.  As a
general rule, the concurrency _should not_ be set higher than the
core-count of the system processing the images.

`--max-concurrency=1` has the default behavior of no concurrency while
processing the files in the directories.  Beginning with 50% of the
cores available is a good start.  Larger numbers can have
diminishing returns as the number of concurrent operations saturate
the file I/O of the system.

Concurrently processing files does have an impact on the order that
messages are written to the console/log and the ability to quickly
terminate the program, as the execution waits for all in-flight
operations to complete before shutting down.

## Development

### Running tests
To run the tests, first install the dev dependencies using

```bash
pip3 install -r requirements-dev.txt
```

Then run the tests using

```bash
pytest
```

To run the tests with coverage reports run
```bash
pytest --cov-report term-missing:skip-covered --cov=src tests/
```

Please add the necessary tests when committing a feature or improvement.


### Pre-commit checks
We leverage the [pre-commit](https://pre-commit.com/) framework to automate some general linting/quality checks.

To install the hooks, from within the activated virtualenv run:

```bash
pre-commit install
```

To manually execute the hooks, run:

```bash

changelog.md
# Changelog
##### `1.13.0`
* Implement `--rmdirs` [#225](https://github.com/ivandokov/phockup/pull/225)
##### `1.12.0`
* Implement `--movedel` [#223](https://github.com/ivandokov/phockup/pull/223)
##### `1.11.3`
* Fixed wrong tag in Git
##### `1.11.2`
* Fixed [deployment to Homebrew](https://github.com/ivandokov/phockup/commit/33aec10e7506fe5f70a8e244e963304ff6f54337)
##### `1.11.1`
* Fixed code styling
##### `1.11.0`
* Added `--from-date` and `--to-datezz options to limit the processed files [#202](https://github.com/ivandokov/phockup/pull/202)
* Merged dependabot PRs
* Improved documentation
##### `1.10.1`
* Fixed python versions for tests and deployments
##### `1.10.0`
* Fixed Dependabot
* Documented AUR install [#175](https://github.com/ivandokov/phockup/pull/175)
* Timezone support [#182](https://github.com/ivandokov/phockup/pull/182)
* Implementation of Issue [#146](https://github.com/ivandokov/phockup/issues/146) to support prefix and suffix [#189](https://github.com/ivandokov/phockup/pull/189)
* Documented `--max-concurrency` [#188](https://github.com/ivandokov/phockup/pull/188)
* Make sure arguments are not split [#190](https://github.com/ivandokov/phockup/pull/190)
* Dependabot PRs
##### `1.9.2`
* Fix brew SHA256 mismatch [#169](https://github.com/ivandokov/phockup/issues/169)
* Fix deploy for buildx [#168](https://github.com/ivandokov/phockup/pull/168)
* Use Python 3.10 for Docker [#167](https://github.com/ivandokov/phockup/pull/167)
* Add dependabot to project [#166](https://github.com/ivandokov/phockup/pull/166)
* Update tests.yml [#164](https://github.com/ivandokov/phockup/pull/164)
* Bump pre-commit packages [#165](https://github.com/ivandokov/phockup/pull/165)
* Fix "WARNING: --use-feature=2020-resolver no longer has any effect..."
##### `1.9.1`
* Specify platforms for Dockerhub action
##### `1.9.0`
* Rename `unknown` folder [#141](https://github.com/ivandokov/phockup/pull/141)
* Week date format [#142](https://github.com/ivandokov/phockup/pull/142)
* [Update snap to core20](https://github.com/ivandokov/phockup/commit/69783c84fe07b94e9b2c62117cf3c0ae5ca2a29e)
* [Fixed missing dep for snap](https://github.com/ivandokov/phockup/commit/b865b56f31c6fde1eadf71540bcf66ceb7744dd3)
##### `1.8.0`
* Added support for threads (`--max-concurrency`) to speed up the process [#123](https://github.com/ivandokov/phockup/pull/123)
##### `1.7.1`
* Fix dependencies due to tqdm [#133](https://github.com/ivandokov/phockup/pull/133)
* Improve check_directories output on error [#132](https://github.com/ivandokov/phockup/pull/132)
* Other improvements [#135](https://github.com/ivandokov/phockup/pull/135), [#128](https://github.com/ivandokov/phockup/pull/128)
##### `1.7.0`
* Add `--progress` functionality [#118](https://github.com/ivandokov/phockup/pull/118)
* Add pre-commit integration [#121](https://github.com/ivandokov/phockup/pull/121)
##### `1.6.5`
* Add missing checkout step to the dockerhub deployment action
##### `1.6.4`
* Add argument "--file-type" to be able to choose between image or video [#114](https://github.com/ivandokov/phockup/issues/114)
* Improved Docker image [#117](https://github.com/ivandokov/phockup/issues/117)
* Automatically deploy new Docker image to Docker Hub [#120](https://github.com/ivandokov/phockup/issues/120)
##### `1.6.3`
* Fixed double `sed`
##### `1.6.2`
* Fixed version extraction for snaps
##### `1.6.1`
* Fixed `--log` argument ([discussion](https://github.com/ivandokov/phockup/pull/106#discussion_r642048830))
* Fixed multiple custom date fields (`-f|--date-field`) [#113](https://github.com/ivandokov/phockup/issues/113)
##### `1.6.0`
* Added `--maxdepth` mode [#104](https://github.com/ivandokov/phockup/issues/104)
* Added `--quiet` mode to hide generic output [#103](https://github.com/ivandokov/phockup/issues/103)
* Fixed tests comatibility for Windows [#102](https://github.com/ivandokov/phockup/issues/102)
* Readme updates
##### `1.5.26`
* Fixed [#98](https://github.com/ivandokov/phockup/issues/98)
* [Disabled automated snap build and deploy for linux/arm64](https://github.com/ivandokov/phockup/issues/99).
##### `1.5.25`
* Fixed [#97](https://github.com/ivandokov/phockup/issues/97)
##### `1.5.24`
* Fixed broken `--date` after the merge of [#87](https://github.com/ivandokov/phockup/issues/87)
##### `1.5.23`
* Removed s930x architecture
* Update snapcraft.yml to more simple setup
##### `1.5.22`
* Fix quotes
##### `1.5.21`
* Fix hard coded variable
##### `1.5.20`
* When taking the current version for snaps get the version for the current architecture and for stable channel
##### `1.5.19`
* Fixed Homebrew sha256 replacing in the deployment workflow job
##### `1.5.18`
* Move Homebrew deployment in a separate job because it was executed multiple times because of the python matrix
##### `1.5.17`
* Fixed workflow wrong step id
##### `1.5.16`
* Fixed snap deploy condition
##### `1.5.15`
* Fixed snap build volume directory
##### `1.5.14`
* Automatic snap deployment thanks to [Daniel Llewellyn's blog post](https://snapcraft.ninja/2020/08/03/snapcraft-continuous-integration-github-actions/)
##### `1.5.13`
* Fix snapcraft version extraction by [#87](https://github.com/ivandokov/phockup/issues/87)
##### `1.5.12`
* Merged [#87](https://github.com/ivandokov/phockup/issues/87)
* Merged [#88](https://github.com/ivandokov/phockup/issues/88)
##### `1.5.11`
* Added Docker support [#75](https://github.com/ivandokov/phockup/issues/75)
##### `1.5.10`
* Merged [#78](https://github.com/ivandokov/phockup/issues/78)
* Merged [#81](https://github.com/ivandokov/phockup/issues/81)
##### `1.5.9`
* Fixed [#70](https://github.com/ivandokov/phockup/issues/70) related to Windows issues
##### `1.5.8`
* Add `--date-field` option to set date extraction fields  [#54](https://github.com/ivandokov/phockup/issues/54)
* Handle regex with optional hour information  [#62](https://github.com/ivandokov/phockup/issues/62)

phockup.py
#!/usr/bin/env python3
import argparse
import logging
import logging.handlers
import os
import re
import sys

from src.date import Date
from src.dependency import check_dependencies
from src.phockup import Phockup

__version__ = '1.13.0'

PROGRAM_DESCRIPTION = """\
Media sorting tool to organize photos and videos from your camera in folders by year, \
month and day.
The software will collect all files from the input directory and copy them to the output
directory without changing the files content. It will only rename the files and  place
them in the proper directory for year, month and day.
"""

DEFAULT_DIR_FORMAT = ['%Y', '%m', '%d']

logger = logging.getLogger('phockup')


def parse_args(args=sys.argv[1:]):
    parser = argparse.ArgumentParser(
        description=PROGRAM_DESCRIPTION,
        formatter_class=argparse.RawTextHelpFormatter)

    parser.version = f"v{__version__}"

    parser.add_argument(
        '-v',
        '--version',
        action='version',
    )

    parser.add_argument(
        '-d',
        '--date',
        action='store',
        type=Date.parse,
        help="""\
Specify date format for OUTPUTDIR directories.

You can choose different year format (e.g. 17 instead of 2017) or decide to skip the
day directories and have all photos sorted in year/month.

Supported formats:
    YYYY - 2016, 2017 ...
    YY   - 16, 17 ...
    MM   - 07, 08, 09 ...
    M    - July, August, September ...
    m    - Jul, Aug, Sept ...
    DD   - 27, 28, 29 ... (day of month)
    DDD  - 123, 158, 365 ... (day of year)
    U    - 00, 01, 53 ... (week of the year, Sunday first day of week)
    W    - 00, 01, 53 ... (week of the year, Monday first day of week)

Example:
    YYYY/MM/DD -> 2011/07/17
    YYYY/M/DD  -> 2011/July/17
    YYYY/m/DD  -> 2011/Jul/17
    YY/m-DD    -> 11/Jul-17
    YYYY/U     -> 2011/30
    YYYY/W     -> 2011/28
""",
    )

    exclusive_group_link_move = parser.add_mutually_exclusive_group()

    exclusive_group_link_move.add_argument(
        '-m',
        '--move',
        action='store_true',
        help="""\
            Instead of copying the process will move all files from the INPUTDIR to the OUTPUTDIR.
            This is useful when working with a big collection of files and the remaining free space
            is not enough to make a copy of the INPUTDIR.
            """,
    )

    exclusive_group_link_move.add_argument(
        '-l',
        '--link',
        action='store_true',
        help="""\
            Instead of copying the process will make hard links to all files in INPUTDIR and place
            them in the OUTPUTDIR.
            This is useful when working with working structure and want to create YYYY/MM/DD
            structure to point to same files.
            """,
    )

    parser.add_argument(
        '-o',
        '--original-names',
        action='store_true',
        help="""\
            Organize the files in selected format or using the default year/month/day format but
            keep original filenames.
            """,
    )

    parser.add_argument(
        '-t',
        '--timestamp',
        action='store_true',
        help="""\
            Use the timestamp of the file (last modified date) if there is no EXIF date information.
            If the user supplies a regex, it will be used if it finds a match in the filename.
            This option is intended as "last resort" since the file modified date may not be
            accurate, nevertheless it can be useful if no other date information can be obtained.
            """,
    )

    parser.add_argument(
        '-y',
        '--dry-run',
        action='store_true',
        help="""\
            Does a trial run with no permanent changes to the filesystem.
            So it will not move any files, just shows which changes would be done.
            """,
    )

    parser.add_argument(
        '-c',
        '--max-concurrency',
        type=int,
        default=1,
        choices=range(1, 255),
        metavar='1-255',
        help="""\
            Sets the level of concurrency for processing files in a directory.
            Defaults to 1. Higher values can improve throughput of file operations
            """,
    )

    parser.add_argument(
        '--maxdepth',
        type=int,
        default=-1,
        choices=range(0, 255),
        metavar='1-255',
        help="""\
            Descend at most 'maxdepth' levels (a non-negative integer) of directories
            """,
    )

    parser.add_argument(
        '-r',
        '--regex',
        action='store',
        type=re.compile,
        help="""\
            Specify date format for date extraction from filenames if there is no EXIF date
            information.

            Example:
                {regex}
                can be used to extract the date from file names like the following
                IMG_27.01.2015-19.20.00.jpg.
            """,
    )

    parser.add_argument(
        '-f',
        '--date-field',
        action='store',
        help="""\
            Use a custom date extracted from the exif field specified.
            To set multiple fields to try in order until finding a valid date, use spaces to
            separate fields inside a string.

            Example:
                DateTimeOriginal
                "DateTimeOriginal CreateDate FileModifyDate"

            These fields are checked by default when this argument is not set:
                "SubSecCreateDate SubSecDateTimeOriginal CreateDate DateTimeOriginal"

            To get all date fields available for a file, do:
                exiftool -time:all -mimetype -j <file>
            """,
    )

    parser.add_argument(
        "--sidecars",
        default=",".join(Phockup.DEFAULT_SIDECAR_EXTENSIONS),
        action="store",
        help="""
Override the set of extensions that are considered to be sidecar files.
Sidecars are files that have the same name as an image or video file, but a
different extension. They typically contain additional metadata pertaining to
the image or video file.

File extensions which are considered to be sidecars if they are named the same
as a corresponding image file are:

    """
        + str(Phockup.DEFAULT_SIDECAR_EXTENSIONS)
        + """

So, for example, if image.jpg exists, then image.xmp (or image.jpg.xmp) will be
considered a sidecar file of image.jpg.

When moving the main file, sidecars will inherit the name of the main file and
be moved to the same location rather than being placed in the unknown
directory.

Using this argument you can change this set. For example, to only consider XMP
and JSON:
    --sidecars='xmp,json'
""",
    )

    exclusive_group_debug_silent = parser.add_mutually_exclusive_group()

    exclusive_group_debug_silent.add_argument(
        '--debug',
        action='store_true',
        default=False,
        help="""\
            Enable debugging.  Alternately, set the LOGLEVEL environment variable to DEBUG
            """,
    )

    exclusive_group_debug_silent.add_argument(
        '--quiet',
        action='store_true',
        default=False,
        help="""\
            Run without output.
            """,
    )

    exclusive_group_debug_silent.add_argument(
        '--progress',
        action='store_true',
        default=False,
        help="""\
            Run with progressbar output.
            """,
    )

    parser.add_argument(
        '--log',
        action='store',
        help="""\
            Specify the output directory where your log file should be exported.
            This flag can be used in conjunction with the flag `--quiet` or `--progress`.
            """,
    )

    parser.add_argument(
        'input_dir',
        metavar='INPUTDIR',
        help="""\
            Specify the source directory where your photos are located.
            """,
    )

    parser.add_argument(
        'output_dir',
        metavar='OUTPUTDIR',
        help="""\
            Specify the output directory where your photos should be exported.
            """,
    )

    parser.add_argument(
        '--file-type',
        type=str,
        choices=['image', 'video'],
        metavar='image|video',
        help="""\
            By default, Phockup addresses both image and video files.
            If you want to restrict your command to either images or
            videos only, use `--file-type=[image|video]`.
            """,
    )

    parser.add_argument(
        '--no-date-dir',
        type=str,
        default=Phockup.DEFAULT_NO_DATE_DIRECTORY,
        help="""\
            Files without EXIF date information are placed in a directory
            named 'unknown' by default.  This option overrides that
            folder name. e.g. --no-date-dir=misc, --no-date-dir="no date"
            """,
    )

    parser.add_argument(
        '--skip-unknown',
        action='store_true',
        default=False,
        help="""\
            Ignore files that don't contain valid EXIF data for the criteria specified.
            This is useful if you intend to make multiple passes over an input directory
            with varying and specific EXIF fields that are note checked by default.
            """,
    )

    parser.add_argument(
        '--movedel',
        action='store_true',
        default=False,
        help="""\
            DELETE source files which are determined to be duplicates of files
            already transferred.  Only valid in conjunction with both `--move`
            and `--skip-unknown`.
            """,
    )

    parser.add_argument(
        '--rmdirs',
        action='store_true',
        default=False,
        help="""\
            DELETE empty directories after processing.  Only valid in
            conjunction with `--move`.
            """,
    )

    parser.add_argument(
        '--output_prefix',
        type=str,
        default='',
        help="""\
            String to prepend to the output directory to aid in sorting
            files by an additional level prior to sorting by date.  This
            string will immediately follow the output path and is intended
            to allow runtime setting of the output path (e.g. via $USER,
            $HOSTNAME, %%USERNAME%%, etc.)
            """,
    )

    parser.add_argument(
        '--output_suffix',
        type=str,
        default='',
        help="""\
            String to append to the destination directory to aid in sorting
            files by an additional level after sorting by date.
            """,
    )

    parser.add_argument(
        '--from-date',
        type=str,
        default=None,
        help="""\
            Limit the operations to the files that are newer than --from-date (inclusive).
            The date must be specified in format YYYY-MM-DD. Files with unknown date won't be skipped.
            """,
    )

    parser.add_argument(
        '--to-date',
        type=str,
        default=None,
        help="""\
            Limit the operations to the files that are older than --to-date (inclusive).
            The date must be specified in format YYYY-MM-DD. Files with unknown date won't be skipped.
            """,
    )

    return parser.parse_args(args)


def setup_logging(options):
    """Configure logging."""
    root = logging.getLogger('')
    root.setLevel(logging.WARNING)
    formatter = logging.Formatter(
        '[%(asctime)s] - [%(levelname)s] - %(message)s', '%Y-%m-%d %H:%M:%S')
    ch = logging.StreamHandler()
    ch.setFormatter(formatter)
    root.addHandler(ch)
    if not options.quiet ^ options.progress:
        logger.setLevel(options.debug and logging.DEBUG or logging.INFO)
    else:
        logger.setLevel(logging.WARNING)
    if options.log:
        logfile = os.path.expanduser(options.log)
        fh = logging.FileHandler(logfile)
        fh.setFormatter(formatter)
        logger.addHandler(fh)
    logger.debug("Debug logging output enabled.")
    logger.debug("Running Phockup version %s", __version__)


def main(options):
    check_dependencies()

    return Phockup(
        options.input_dir,
        options.output_dir,
        dir_format=options.date,
        move=options.move,
        link=options.link,
        date_regex=options.regex,
        original_filenames=options.original_names,
        timestamp=options.timestamp,
        date_field=options.date_field,
        dry_run=options.dry_run,
        quiet=options.quiet,
        progress=options.progress,
        max_depth=options.maxdepth,
        file_type=options.file_type,
        max_concurrency=options.max_concurrency,
        no_date_dir=options.no_date_dir,
        skip_unknown=options.skip_unknown,
        movedel=options.movedel,
        rmdirs=options.rmdirs,
        output_prefix=options.output_prefix,
        output_suffix=options.output_suffix,
        from_date=options.from_date,
        to_date=options.to_date
    )


if __name__ == '__main__':
    try:
        options = parse_args()
        setup_logging(options)
        main(options)
    except Exception as e:
        logger.warning(e)
        sys.exit(1)
    except KeyboardInterrupt:
        logger.error("Exiting phockup...")
        sys.exit(1)
    sys.exit(0)
